File: lib/src/models/page_layout.dart
import 'package:meta/meta.dart';
import 'package:collection/collection.dart';

@immutable
class PageMargins {
  final String? type; // "odd", "even", "both"
  final double? leftMargin;
  final double? rightMargin;
  final double? topMargin;
  final double? bottomMargin;

  const PageMargins({
    this.type,
    this.leftMargin,
    this.rightMargin,
    this.topMargin,
    this.bottomMargin,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is PageMargins &&
          runtimeType == other.runtimeType &&
          type == other.type &&
          leftMargin == other.leftMargin &&
          rightMargin == other.rightMargin &&
          topMargin == other.topMargin &&
          bottomMargin == other.bottomMargin;

  @override
  int get hashCode =>
      type.hashCode ^
      leftMargin.hashCode ^
      rightMargin.hashCode ^
      topMargin.hashCode ^
      bottomMargin.hashCode;

  @override
  String toString() {
    return 'PageMargins{type: $type, leftMargin: $leftMargin, rightMargin: $rightMargin, topMargin: $topMargin, bottomMargin: $bottomMargin}';
  }
}

@immutable
class PageLayout {
  final double? pageHeight;
  final double? pageWidth;
  final List<PageMargins> pageMargins; // Can have up to 2 (odd/even) or one for "both"

  const PageLayout({
    this.pageHeight,
    this.pageWidth,
    this.pageMargins = const [],
  });

   @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is PageLayout &&
          runtimeType == other.runtimeType &&
          pageHeight == other.pageHeight &&
          pageWidth == other.pageWidth &&
          const DeepCollectionEquality().equals(pageMargins, other.pageMargins);

  @override
  int get hashCode =>
      pageHeight.hashCode ^
      pageWidth.hashCode ^
      const DeepCollectionEquality().hash(pageMargins);

  @override
  String toString() {
    return 'PageLayout{pageHeight: $pageHeight, pageWidth: $pageWidth, pageMargins: $pageMargins}';
  }
}

/// Represents scaling information in a MusicXML document.
/// Typically found within <defaults>.
@immutable
class Scaling {
  /// The number of millimeters per unit.
  final double millimeters;

  /// The number of tenths per unit.
  final double tenths;

  /// Creates a new [Scaling] instance.
  const Scaling({
    required this.millimeters,
    required this.tenths,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Scaling &&
          runtimeType == other.runtimeType &&
          millimeters == other.millimeters &&
          tenths == other.tenths;

  @override
  int get hashCode => millimeters.hashCode ^ tenths.hashCode;

  @override
  String toString() {
    return 'Scaling{millimeters: $millimeters, tenths: $tenths}';
  }
}

File: MusicXMLParser/Models/PageLayout.cs
using System.Collections.Generic;
using System.Linq;

namespace MusicXMLParser.Models
{
    public class PageMargins
    {
        public string Type { get; set; } // "odd", "even", "both"
        public double? LeftMargin { get; set; }
        public double? RightMargin { get; set; }
        public double? TopMargin { get; set; }
        public double? BottomMargin { get; set; }

        public override bool Equals(object obj)
        {
            if (obj is PageMargins other)
            {
                return Type == other.Type &&
                       LeftMargin == other.LeftMargin &&
                       RightMargin == other.RightMargin &&
                       TopMargin == other.TopMargin &&
                       BottomMargin == other.BottomMargin;
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Type, LeftMargin, RightMargin, TopMargin, BottomMargin);
        }

        public override string ToString()
        {
            return $"PageMargins{{Type: {Type}, LeftMargin: {LeftMargin}, RightMargin: {RightMargin}, TopMargin: {TopMargin}, BottomMargin: {BottomMargin}}}";
        }
    }

    public class PageLayout
    {
        public double? PageHeight { get; set; }
        public double? PageWidth { get; set; }
        public List<PageMargins> PageMargins { get; set; } = new List<PageMargins>(); // Can have up to 2 (odd/even) or one for "both"

        public override bool Equals(object obj)
        {
            if (obj is PageLayout other)
            {
                return PageHeight == other.PageHeight &&
                       PageWidth == other.PageWidth &&
                       PageMargins.SequenceEqual(other.PageMargins);
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(PageHeight, PageWidth, PageMargins);
        }

        public override string ToString()
        {
            return $"PageLayout{{PageHeight: {PageHeight}, PageWidth: {PageWidth}, PageMargins: {string.Join(", ", PageMargins)}}}";
        }
    }

    /// <summary>
    /// Represents scaling information in a MusicXML document.
    /// Typically found within <defaults>.
    /// </summary>
    public class Scaling
    {
        /// <summary>
        /// The number of millimeters per unit.
        /// </summary>
        public double Millimeters { get; set; }

        /// <summary>
        /// The number of tenths per unit.
        /// </summary>
        public double Tenths { get; set; }

        public Scaling(double millimeters, double tenths)
        {
            Millimeters = millimeters;
            Tenths = tenths;
        }

        public override bool Equals(object obj)
        {
            if (obj is Scaling other)
            {
                return Millimeters == other.Millimeters &&
                       Tenths == other.Tenths;
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Millimeters, Tenths);
        }

        public override string ToString()
        {
            return $"Scaling{{Millimeters: {Millimeters}, Tenths: {Tenths}}}";
        }
    }
}
